stl/inc/xutility

```diff
@@ -2668,9 +2744,223 @@ namespace ranges {
     // CONCEPT ranges::contiguous_range
     template <class _Rng>
     concept contiguous_range = range<_Rng> && contiguous_iterator<iterator_t<_Rng>> && requires(_Rng& __r) {
-        { data(__r) } -> same_as<add_pointer_t<range_reference_t<_Rng>>>;
+        { _RANGES data(__r) } -> same_as<add_pointer_t<range_reference_t<_Rng>>>;
     };
     // clang-format on
+
+    // CLASS ranges::_Not_quite_object
+    class _Not_quite_object {
+    public:
+        // Some overload sets in the library have the property that their constituent function templates are not visible
+        // to argument-dependent name lookup (ADL) and that they inhibit ADL when found via unqualified name lookup.
+        // This property allows these overload sets to be implemented as function objects. We derive such function
+        // objects from this type to remove some typical object-ish behaviors which helps users avoid depending on their
+        // non-specified object-ness.
+
+        struct _Construct_tag {};
+
+        _Not_quite_object() = delete;
+
+        constexpr explicit _Not_quite_object(_Construct_tag) noexcept {}
+
+        _Not_quite_object(const _Not_quite_object&) = delete;
+        _Not_quite_object& operator=(const _Not_quite_object&) = delete;
+
+        void operator&() const = delete;
+
+    protected:
+        ~_Not_quite_object() = default;
+    };
+
+    // CLASS ranges::_Advance_fn
+    class _Advance_fn : private _Not_quite_object {
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        template <input_or_output_iterator _It>
+        constexpr void operator()(_It& _Where, iter_difference_t<_It> _Off) const {
+            if constexpr (random_access_iterator<_It>) {
+                _Where += _Off;
+            } else {
+                if constexpr (!bidirectional_iterator<_It>) {
+                    _STL_ASSERT(_Off >= 0, "negative advance of non-bidirectional iterator");
+                }
+
+                auto&& _UWhere              = _Get_unwrapped_n(_STD move(_Where), _Off);
+                constexpr bool _Need_rewrap = !is_reference_v<decltype(_Get_unwrapped_n(_STD move(_Where), _Off))>;

```

Comment:
Other occurences use `static_cast<_It&&>` rather than `std::move`. As far as I can tell both are the same so one should be consistent

Reply:
We don't have consensus on using `static_cast` uniformly everywhere instead of `_STD move` and `_STD forward`. We don't agree that the throughput benefits of `static_cast` are great enough to justify the readability hit in general. We *have* established a consensus for preferring `static_cast` in metaprogramming (traits, SFINAE, concepts, `constexpr` functions called by traits/SFINAE/concepts) that is evaluated as part of overload resolution to reduce overhead from instantiation of templates that may not end up being called.

I won't bet my life that there aren't usages that contradict these guidelines, but this `move` is consistent.

Reply:
For normal programming (not metaprogramming), I am strongly of the opinion that Microsoft-internal VSO-437566 "`<utility>` C1XX should understand move and forward, and not emit functions into the obj/PDB" should be implemented, so we can write natural code.

Proliferating `static_cast` will desensitize us to casts.